### PR TITLE
Parametrize PeriodIndex tests

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexes/period/test_arithmetic.py [hunk 1]
```diff
@@ -335,102 +272,228 @@ def test_sub_isub(self):
...
+        Timedelta(days=365)])
+    def test_pi_add_iadd_timedeltalike_freq_mismatch_monthly(self, other):
+        rng = pd.period_range('2014-01', '2016-12', freq='M')
+        msg = 'Input has different freq(=.+)? from PeriodIndex\\(freq=M\\)'
+        with tm.assert_raises_regex(period.IncompatibleFrequency, msg):
+            rng + other
+        with tm.assert_raises_regex(period.IncompatibleFrequency, msg):
+            rng += other
+
+    @pytest.mark.parametrize('other', [
```
Comment (jreback) [inline pandas/tests/indexes/period/test_arithmetic.py]: so I'l like to see you not repeat the parameterization on every test. you can do this in a number of ways

- a fixture
- make a list of the parameters (then its easy to add on a particular test)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/period/test_arithmetic.py [hunk 2]
```diff
@@ -11,7 +11,46 @@
 import pandas.core.indexes.period as period
 
 
+_common_mismatch = [pd.offsets.YearBegin(2),
+                    pd.offsets.MonthBegin(1),
+                    pd.offsets.Minute()]
+
+not_hourly_offsets = [timedelta(minutes=30),
```
Comment (jreback) [inline pandas/tests/indexes/period/test_arithmetic.py]: from here down should be fixtures, then you don't need the parameterize on each function, just pass the argument.

File: pandas/tests/indexes/period/test_astype.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        tm.assert_numpy_array_equal(idx.astype(object).values, exp)
+        tm.assert_numpy_array_equal(idx._mpl_repr(), exp)
+
+        exp = np.array([pd.Period('2011-01-01', freq='D'), pd.NaT],
+                       dtype=object)
+        idx = pd.PeriodIndex(['2011-01-01', pd.NaT], freq='D')
+        tm.assert_numpy_array_equal(idx.astype(object).values, exp)
+        tm.assert_numpy_array_equal(idx._mpl_repr(), exp)
+
+    # TODO: de-duplicate this version (from test_ops) with the one above
```
Comment (jreback) [inline pandas/tests/indexes/period/test_astype.py]: when deduping (can be next pass). see if anyting can be parameterized

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/period/test_arithmetic.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/period/test_arithmetic.py]: just pushed with this change.

**Comment by jbrockmendel** [comment]
Multiple Travis builds have been running for almost 2hours. Is this the only PR affected?

File: pandas/tests/indexes/period/test_arithmetic.py [hunk 4]
```diff
@@ -11,7 +11,63 @@
...
 
+_common_mismatch = [pd.offsets.YearBegin(2),
+                    pd.offsets.MonthBegin(1),
+                    pd.offsets.Minute()]
+
+
+@pytest.fixture(params=[timedelta(minutes=30),
+                        np.timedelta64(30, 's'),
+                        Timedelta(seconds=30)] + _common_mismatch)
+def not_hourly(request):
```
Comment (TomAugspurger) [inline pandas/tests/indexes/period/test_arithmetic.py]: Could I bother you for docstrings on these? They're printed out when running `pytest --fixtures`, which I find really helpful in understanding what fixtures we already provide.

Comment (jbrockmendel) [reply pandas/tests/indexes/period/test_arithmetic.py]: Just pushed a commit with these added.  Let me know if they need to be fleshed out etc.

If you find the motivation, I've got a handful of small PRs that need a nudge to get over the line (i.e. haven't gotten the "ping on green", but plausibly could): #19582, #19650, #19692, #19365, #19611, #19661.  Less trivial but important from a blocker-for-bugfixes standpoint is #19672.

File: pandas/tests/indexes/period/test_astype.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+import numpy as np
+import pytest
+
+import pandas as pd
+import pandas.util.testing as tm
+from pandas import NaT, Period, PeriodIndex, Int64Index, Index, period_range
+
+
+class TestPeriodIndexAsType(object):
```
Comment (jreback) [inline pandas/tests/indexes/period/test_astype.py]: can you remove  the class. normally don't use a class unless you have 2 (and then I would question why we are not doing in sub-modules), unless very short

File: pandas/tests/indexes/period/test_arithmetic.py (hunk 4 above)
Comment (jreback) [reply pandas/tests/indexes/period/test_arithmetic.py]: woa didn't realize fixtures prints these out. let' fix these next for all ones that don't have them. il'll create an issue

Comment (jreback) [reply pandas/tests/indexes/period/test_arithmetic.py]: https://github.com/pandas-dev/pandas/issues/19731

File: pandas/tests/indexes/period/test_arithmetic.py [hunk 6]
```diff
@@ -305,16 +300,24 @@ def test_pi_add_int(self, one):
         rng += one
         tm.assert_index_equal(rng, expected)
 
+    def test_pi_sub_isub_int(self, one):
```
Comment (jreback) [inline pandas/tests/indexes/period/test_arithmetic.py]: prob want to doc-string one (maybe rename to slightly more informative)

**Review by jreback — CHANGES_REQUESTED** [review]
small changes. mainly interested in the de-classing in astype. other things can also be next pass. ping on green.

File: pandas/tests/indexes/period/test_astype.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/period/test_astype.py]: I tend to agree on the why-is-a-class-needed bit.  I'd also like to keep symmetry with the test_astype modules in tests.indexes.datetimes and tests.indexes.timedeltas.  In an upcoming pass I'll de-class all three at the same time.

**Comment by jbrockmendel** [comment]
ping

**Comment by TomAugspurger** [comment]
Thanks!

**Pull request merged**